### PR TITLE
support: add switchover for redis in kbagent

### PR DESCRIPTION
**Pull request by DWJ-Squirtle**

### CODE REVIEW

File: addons/redis/scripts/redis-switchover.sh [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    else 
+        current_replica_priority=$(redis_config_get "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$redis_get_cmd" | sed -n '2p')
+    fi
+
+    redis_set_recover_cmd="CONFIG SET replica-priority $current_replica_priority"
+
+    if ! is_empty "$REDIS_DEFAULT_PASSWORD"; then
+        call_func_with_retry 3 5 execute_sub_command "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$REDIS_DEFAULT_PASSWORD" "$redis_set_switchover_cmd" || exit 1
+    else
+        call_func_with_retry 3 5 execute_sub_command "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$redis_set_switchover_cmd" || exit 1
```
Comment (zjx20) [inline addons/redis/scripts/redis-switchover.sh:141]: ditto

File: addons/redis/scripts/redis-switchover.sh [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    fi
+    set_xtrace_when_ut_mode_false
+    # TODO: check the role in kernel before switchover
+    old_ifs="$IFS"
+    IFS=','
+    set -f
+    read -ra sentinel_pod_fqdn_list <<< "${SENTINEL_POD_FQDN_LIST}"
+    set +f
+    IFS="$old_ifs"
```
Comment (zjx20) [inline addons/redis/scripts/redis-switchover.sh:150]: Simply the code:
```bash
    old_ifs="$IFS"
    IFS=','
    set -f
    read -ra sentinel_pod_fqdn_list <<< "${SENTINEL_POD_FQDN_LIST}"
    set +f
    IFS="$old_ifs"
```

To:

```bash
    IFS=',' read -ra sentinel_pod_fqdn_list <<< "${SENTINEL_POD_FQDN_LIST}"
```

File: addons/redis/scripts/redis-switchover.sh [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    if ! is_empty "$REDIS_DEFAULT_PASSWORD"; then
+        call_func_with_retry 3 5 execute_sub_command "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$REDIS_DEFAULT_PASSWORD" "$redis_set_recover_cmd" || exit 1
+    else
+        call_func_with_retry 3 5 execute_sub_command "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$redis_set_recover_cmd" || exit 1
+    fi
+}
+
+switchoverWithoutCandidate() {
+    # TODO: check the role in kernel before switchover
+    old_ifs="$IFS"
```
Comment (zjx20) [inline addons/redis/scripts/redis-switchover.sh:176]: ditto

File: addons/redis/scripts/redis-switchover.sh [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    fi
+
+    redis_set_recover_cmd="CONFIG SET replica-priority $current_replica_priority"
+
+    if ! is_empty "$REDIS_DEFAULT_PASSWORD"; then
+        call_func_with_retry 3 5 execute_sub_command "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$REDIS_DEFAULT_PASSWORD" "$redis_set_switchover_cmd" || exit 1
+    else
+        call_func_with_retry 3 5 execute_sub_command "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$redis_set_switchover_cmd" || exit 1
+    fi
+    set_xtrace_when_ut_mode_false
```
Comment (zjx20) [inline addons/redis/scripts/redis-switchover.sh:143]: Don't enable tracing, to hide the following $SENTINEL_PASSWORD and $REDIS_DEFAULT_PASSWORD from stdout.

File: addons/redis/scripts/redis-switchover.sh [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            break
+        fi
+    done
+
+    if [ "$success" = false ]; then
+        echo "All Sentinel failover attempts failed."
+        exit 1
+    fi
+    # TODO: check switchover result
+    if ! is_empty "$REDIS_DEFAULT_PASSWORD"; then
```
Comment (zjx20) [inline addons/redis/scripts/redis-switchover.sh:167]: ditto

File: addons/redis/templates/cmpd-redis-7.yaml [hunk 6]
```diff
@@ -260,6 +266,15 @@ spec:
...
           - "redis-cli -h $(KB_ACCOUNT_ENDPOINT) -a $REDIS_DEFAULT_PASSWORD $(KB_ACCOUNT_STATEMENT) && redis-cli -h $(KB_ACCOUNT_ENDPOINT) -a $REDIS_DEFAULT_PASSWORD acl save "
+    switchover:
+      exec:
+        image: {{ include "redis7.image" . }}
+        container: redis
+        command:
+          - /bin/bash
+          - -c
+          - |
+            {{- .Files.Get "scripts/redis-switchover.sh" | nindent 14 }}
```
Comment (zjx20) [inline addons/redis/templates/cmpd-redis-7.yaml:277]: nindent 12

**Review by zjx20 — APPROVED** [review]

File: addons/redis/templates/cmpd-redis-7.yaml [hunk 7]
```diff
@@ -223,6 +223,12 @@ spec:
           - /tools/config/dbctl/components
           - redis
           - getrole
+        env:
+          - name: KB_POD_NAME
```
Comment (Y-Rookie) [inline addons/redis/templates/cmpd-redis-7.yaml:227]: rename `KB_POD_NAME` to `CURRENT_POD_NAME` or other name without “KB_” prefix

File: addons/redis/templates/cmpd-redis-7.yaml [hunk 8]
```diff
@@ -260,6 +266,15 @@ spec:
...
           - "redis-cli -h $(KB_ACCOUNT_ENDPOINT) -a $REDIS_DEFAULT_PASSWORD $(KB_ACCOUNT_STATEMENT) && redis-cli -h $(KB_ACCOUNT_ENDPOINT) -a $REDIS_DEFAULT_PASSWORD acl save "
+    switchover:
+      exec:
+        image: {{ include "redis7.image" . }}
+        container: redis
+        command:
+          - /bin/bash
+          - -c
+          - |
+            {{- .Files.Get "scripts/redis-switchover.sh" | nindent 12 }}
```
Comment (Y-Rookie) [inline addons/redis/templates/cmpd-redis-7.yaml:277]: do not use `.Files.Get`

File: addons/redis/scripts/redis-switchover.sh [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        fi
+    done
+
+    if [ "$success" = false ]; then
+        echo "All Sentinel failover attempts failed."
+        exit 1
+    fi
+    # TODO: check switchover result
+    call_func_with_retry 3 5 execute_sub_command "$KB_SWITCHOVER_CANDIDATE_FQDN" "6379" "$REDIS_DEFAULT_PASSWORD" "$redis_set_recover_cmd" || exit 1
+    set_xtrace_when_ut_mode_false
```
Comment (Y-Rookie) [inline addons/redis/scripts/redis-switchover.sh:148]: set back the `replica-priority` after switchover?

File: addons/redis/scripts/redis-switchover.sh [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        return 1
+    fi
+
+    echo "$output"
+    return 0
+}
+
+switchoverWithCandidate() {
+    redis_get_cmd="CONFIG GET replica-priority"
+    redis_set_switchover_cmd="CONFIG SET replica-priority 1"
```
Comment (Y-Rookie) [inline addons/redis/scripts/redis-switchover.sh:122]: how about the others secondaries also have the same replica-priority=1?

**Review by Y-Rookie — APPROVED** [review]

**Review by zjx20 — APPROVED** [review]

**Pull request merged**